### PR TITLE
use lazyComponent wrapper for React lazy-loaded components

### PR DESCRIPTION
**Pull request by sqs**
This removes some boilerplate and will make it easier to support hot reloading for React components that are lazily loaded (for Webpack code splitting).

```
codemod -m --extensions tsx,ts -d web/src "React.lazy\(async \(\) => \(\{\s*default: \(await import\('([^']+)'\)\)\s*\.(\w+),?\s*\}\)\)" 'lazyComponent(() => import("\1"), "\2")'
```

Also:

- Removes unnecessary `const`s at the top of the file where possible.
- make lazy-loading of MonacoSettingsEditor more robust. Hot reloading makes it so that the reference to MonacoSettingsEditor is not the class, but rather a proxy component. This means that isStandaloneCodeEditor and addEditorAction aren't defined as methods on the value when hot reloading is enabled. To work around this issue, this commit uses a more compatible way of lazy-loading the class.

### CODE REVIEW

File: web/src/util/asyncComponent.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * Returns a lazy-loaded reference to a React component in another module.
+ *
+ * This should be used in URL routes and anywhere else that Webpack code splitting can occur, to
+ * avoid all referenced components being in the initial bundle.
+ *
+ * @param componentFactory Asynchronously imports the component's module; e.g., `() =>
+ * import('./MyComponent')`.
+ * @param name The export binding name of the component in its module.
+ */
+export const asyncComponent = <P extends {}, K extends string>(
```
Comment (felixfbecker) [inline web/src/util/asyncComponent.ts]: This seems like a misnomer to me, lots of components could be considered "async", but these are simply lazy loaded. How about `lazyLoadComponent()` or similar?

File: web/src/util/asyncComponent.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    componentFactory: () => Promise<{ [k in K]: React.ComponentType<P> }>,
+    name: K
+) => {
+    // Force returning a React.FunctionComponent-like so our result is callable (because it's used
+    // in <Route render={...} /> elements where it is expected to be callable).
+    const lazyComponent = React.lazy(async () => {
+        const component: React.ComponentType<P> = (await componentFactory())[name]
+        return { default: component }
+    })
+    return (props: PropsWithRef<PropsWithChildren<P>> & Attributes) => React.createElement(lazyComponent, props)
```
Comment (felixfbecker) [inline web/src/util/asyncComponent.ts]: Isn't this equivalent to `<LazyComponent {...props} />`?

File: web/src/util/asyncComponent.ts (hunk 1 above)
Comment (sqs) [reply web/src/util/asyncComponent.ts]: Sounds good. Will rename to lazyComponent (to cut a few chars and keep it shorter).

File: web/src/util/asyncComponent.ts (hunk 2 above)
Comment (sqs) [reply web/src/util/asyncComponent.ts]: Yes, good point. Will push a change.

**Comment by sqs** [comment]
@felixfbecker PTAL. I've updated it per your comments.

**Comment by sqs** [comment]
@felixfbecker Good idea. Filed https://github.com/sourcegraph/sourcegraph/issues/4165 for me to follow up on that.

**Comment by felixfbecker** [comment]
One thing that is unfortunate about this change is that now references to components from routes are not statically known anymore and it is not possible to find where a component is used in a route definition. They will always show "0 references":

image

Do you have an idea to improve this? Is there any disadvantage to doing `import(...).then(m => m.Component)` instead?

**Comment by sqs** [comment]
@felixfbecker I forget if there was a reason to not do that. If it works, then definitely do it.

**Pull request merged**